### PR TITLE
New argument `compute_on_cpu`

### LINKED ISSUE #848: Allow caching on cpu when compute some metrics.
## 🚀 Feature

Currently, some metrics like FID and KID only use GPU computing.
Since all datasets are cached in gpu memory, there is a big limitation on hardware.
It may be helpful to add `compute_on_cpu` option to compute these metrics like below.

```python
def update(self, imgs: Tensor, real: bool) -> None:  # type: ignore
    """Update the state with extracted features.

    Args:
        imgs: tensor with images feed to the feature extractor
        real: bool indicating if imgs belong to the real or the fake distribution
    """
    features = self.inception(imgs)

    if self.compute_on_cpu:
        features = features.detach().cpu()

    if real:
        self.real_features.append(features)
    else:
        self.fake_features.append(features)
```

### Motivation

I can compute FID on cloud system but not my local, because of the memory issue.
Currently, warn large memory foorprint. #468

### LINKED ISSUE #866: `MeanAveragePrecision` CPU overload when using multi-device ddp
## 🐛 Bug

I was using `pytorchlightning.Trainer` in `ddp` mode with 8 GPUs on a HPC. There are 5k images in my validation set, and I find all the CPU cores are fully loaded (100%) for extremely long time (~2 hours) at `validation_epoch_end()` stage.

### To Reproduce

1. using `ddp` mode with `number of device > 2`
2. using large validation/test set.

#### Code sample

```python
class Model(pl.LightningModule):
    def __init__(self):
        self.det = FasterRCNN(...)
        self.map = MeanAveragePrecision(compute_on_step=False)
        ...

    ...

    def validation_step(self, batch, batch_idx):
        images, target = batch
        preds = self.det(images)
        self.map.update(preds=preds, target=target)

    def validation_epoch_end(self, outs):
        val_map_out = self.map.compute()  <-- CPU overload
        self.map.reset()
```

### Expected behavior

The same computation takes place in `N` separate sub-process simultaneously. If the computations are on GPU, they should finish in roughly the same time. But current implementation moves all tensors to CPU and CPU overload occurs while `N` increases.

- ddp with 1 GPU, number of fully loaded GPU cores 16/32, takes ~6 mins.
- ddp with 2 GPU, number of fully loaded GPU cores 32/32, takes ~6 mins.
- ddp with >2 GPU, number of fully loaded GPU cores 32/32, takes > 1 hours.

### Environment

- OS (e.g., Linux): Linux
- Python: 3.8
- PyTorch: 1.8.1 and 1.10.2
- pytorch-lightning: 1.5.10
- torchmetrics: 0.7.2

### Temp solution

My temp solution is re-enable the mAP computation on GPU, although the GPU version is slower than CPU version #677 

I commented the following lines and made other relevant variables on the correct device.

```python
    def compute(self) -> dict:
        # move everything to CPU, as we are faster here
        # self.detection_boxes = [box.cpu() for box in self.detection_boxes]
        # self.detection_labels = [label.cpu() for label in self.detection_labels]
        # self.detection_scores = [score.cpu() for score in self.detection_scores]
        # self.groundtruth_boxes = [box.cpu() for box in self.groundtruth_boxes]
        # self.groundtruth_labels = [label.cpu() for label in self.groundtruth_labels]
```

Should we make `compute_on_cpu` optional? and any suggestion according to TorchMetrics' API design?

I'm also keeping my eyes on the work of improving mAP performance on both CPU and GPU. #742

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Rework of #851 
Fixes #848
General solution that allows *all* metrics where the metric state is a list to store the state on CPU instead of GPU through a new keyword argument `compute_on_cpu=True/False` (default `False`). This can beneficial in situations where the user do not have that much VRAM. Consequence of enabling this feature is that code in `compute` will be executed on CPU (code in `update` will still be executed on GPU).

Fixes #866
Related as the MAP metric currently moves all metric states to CPU for faster computations. As the issue points out this can overload the CPU. Feature in this PR essentially also fixes this by making it an option to move to CPU by setting `MeanAveragePrecision(compute_on_cpu=True)`.

## Before submitting

- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by maximsch2** [comment]
A few questions:
- If I'm reading code correctly, `self.device` will still be `gpu` with this during `compute`? I'm wondering if this will break some metrics. 
- The implicit assumption here is that a List state is only appended during `update` as otherwise we'll get device mismatches, right?
- This doesn't help with other metrics that can occupy significant memory (e.g. binned curve metrics) on GPU as those are allocated as fixed size tensors. In that case though, I guess, a user can manually place them on CPU and transfer data to CPU before `update`. Is that what we would be recommending people?

**Comment by justusschock** [comment]
@maximsch2 I think the device should indicate the input and output device of a metric. E.g. even if the metic states are on cpu, it would require the devices of the inputs to be cuda and handle the device-transfer logic accordingly and the output would also be on gpu.

**Comment by Borda** [comment]
@SkafteNicki mind resolve conflicts :rabbit: 
@justusschock mind review? :fish:

**Comment by Borda** [comment]
@SkafteNicki @justusschock pls ^^ :rabbit:

**Review by justusschock — APPROVED** [review]

**Comment by Borda** [comment]
seems there is a DDP issue... :/
